### PR TITLE
✨ Add `ReconcileStrategy` field to support continuous reconciliation or installing the Helm chart once

### LINKED ISSUE #65: Support one-shot mode
**User Story**

As a user/operator I would like to be able to use this provider to bootstrap addons on a workload cluster, but then allow the workload cluster to take over their management. This mode is currently supported for CRS and I think there are use cases for mirroring it here.

**Detailed Description**

Add a field to the HelmChartProxy to request a "one-shot" mode for the helm release. If this is turned on, the provider requeues the helm chart installation until it gets a successful install on the workload cluster and then forgets about the helm release. After a successful install, there should be no more reconciliation, drift-detection, deletions or anything happening on that helm release irrespectively of what happens on the management cluster or on the workload cluster.

The use case for this is bootstrapping components that are crucial for successful initialization of a cluster (CNI, cloud-controller, some workload-cluster package manager itself, etc...) but then delegating the subsequent management of such releases to the workload cluster itself (which might have its own package manager or its own lifecycle manager for helm releases).

/kind feature

### PR DESCRIPTION
**Pull request by Jont828**
**What this PR does / why we need it**: Introduces a field called `ReconcileStrategy` which allows the user to specify if reconciliation will be continuous (as usual) or if the Helm chart should be successfully installed and no longer reconciled. This is intended as a "fire-and-forget" mode where users intend to hand off their Helm chart to GitOps or to make out of band changes.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #65

### CODE REVIEW

File: controllers/helmchartproxy/helmchartproxy_controller.go [hunk 1]
```diff
@@ -188,10 +188,14 @@ func (r *HelmChartProxyReconciler) reconcileNormal(ctx context.Context, helmChar
 	log := ctrl.LoggerFrom(ctx)
 
 	log.V(2).Info("Starting reconcileNormal for chart proxy", "name", helmChartProxy.Name)
+	log.Info("ReconcileStrategy is", "strategy", helmChartProxy.Spec.ReconcileStrategy)
```
Comment (mboersma) [inline controllers/helmchartproxy/helmchartproxy_controller.go:191]: Should this be `log.V(2).Info` like the previous line?

File: controllers/helmchartproxy/helmchartproxy_controller_phases.go [hunk 2]
```diff
@@ -70,22 +70,29 @@ func (r *HelmChartProxyReconciler) reconcileForCluster(ctx context.Context, helm
...
 	}
-	// log.V(2).Info("Found existing HelmReleaseProxy", "cluster", cluster.Name, "release", existingHelmReleaseProxy.Name)
 
-	if existingHelmReleaseProxy != nil && shouldReinstallHelmRelease(ctx, existingHelmReleaseProxy, helmChartProxy) {
-		log.V(2).Info("Reinstalling Helm release by deleting and creating HelmReleaseProxy", "helmReleaseProxy", existingHelmReleaseProxy.Name)
-		if err := r.deleteHelmReleaseProxy(ctx, existingHelmReleaseProxy); err != nil {
-			conditions.MarkFalse(helmChartProxy, addonsv1alpha1.HelmReleaseProxySpecsUpToDateCondition, addonsv1alpha1.HelmReleaseProxyDeletionFailedReason, clusterv1.ConditionSeverityError, "%s", err.Error())
+	if helmChartProxy.Spec.ReconcileStrategy == string(addonsv1alpha1.ReconcileStrategyInstallOnce) {
+		if internal.HasHelmReleaseBeenSuccessfullyInstalled(existingHelmReleaseProxy) {
+			log.Info("HelmReleaseProxy has been installed on InstallOnce mode, nothing to do", "helmReleaseProxy", existingHelmReleaseProxy.Name, "cluster", cluster.Name)
```
Comment (mboersma) [inline controllers/helmchartproxy/helmchartproxy_controller_phases.go:76]: This line too--just confirming that we want `log.Info` without an explicit log level.

File: main.go [hunk 3]
```diff
@@ -108,7 +108,6 @@ func InitFlags(fs *pflag.FlagSet) {
 
 	fs.DurationVar(&syncPeriod, "sync-period", 10*time.Minute,
 		"Minimum interval at which watched resources are reconciled (e.g. 15m)")
-
```
Comment (mboersma) [inline main.go:111]: Maybe restore this blank line for consistency?

File: test/e2e/helm_out_of_band.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	// Get existing HCP from management Cluster
+	existing := &addonsv1alpha1.HelmChartProxy{}
+	key := types.NamespacedName{
+		Namespace: input.HelmChartProxy.Namespace,
+		Name:      input.HelmChartProxy.Name,
+	}
+	err = mgmtClient.Get(ctx, key, existing)
+	Expect(err).NotTo(HaveOccurred())
+
+	// Get the inital Helm release through the HelmChartProxy
```
Comment (mboersma) [inline test/e2e/helm_out_of_band.go:74]: Seems like this comment got detached from its code?

File: test/e2e/helm_out_of_band.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	// Wait for Helm release on workload cluster to have status = deployed
+	existingReleaseWaitInput := GetWaitForHelmReleaseDeployedInput(ctx, workloadClusterProxy, hrpWaitInput.HelmReleaseProxy.Spec.ReleaseName, hrpWaitInput.HelmReleaseProxy.Spec.ReleaseNamespace, specName)
+	existingRelease := WaitForHelmReleaseDeployed(ctx, existingReleaseWaitInput, e2eConfig.GetIntervals(specName, "wait-helm-existingRelease-deployed")...)
+
+	// Invoke UpgradeHelmChart logic here
+	options := &HelmOptions{
+		Values: input.Values,
+	}
+	UpgradeHelmChart(ctx, workloadClusterProxy, existingRelease.Namespace, input.HelmChartProxy.Spec.RepoURL, existingRelease.Chart.Name(), existingRelease.Name, options, existingRelease.Chart.Metadata.Version)
+	// TODO: how do we figure out that the Helm release won't change? Trigger reconcililation maybe?
```
Comment (mboersma) [inline test/e2e/helm_out_of_band.go:94]: Is this `// TODO` something we should figure out before merge, or is it ok to go with just a sleep period?

File: test/e2e/helm_test.go [hunk 6]
```diff
@@ -218,6 +220,154 @@ var _ = Describe("Workload cluster creation", func() {
...
+						ClusterName:           clusterName,
+						HelmChartProxy:        hcp,
+						Values:                []string{"--controller.name=new-nginx-controller-4"},
+						WaitPeriod:            installOnceWaitPeriod,
+					}
+				})
+			})
+
+			// Modify Helm release and expect it to remain unchanged
+			By("Upgrading Helm release out-of-band fifth time", func() {
```
Comment (mboersma) [inline test/e2e/helm_test.go:339]: Any particular reason why we upgraded the Helm release five times (instead of just once)? Maybe we could make a loop out of these five `// Modify Helm release` sections.

File: test/e2e/helm_unchanged.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	// Get existing HCP from management Cluster
+	existing := &addonsv1alpha1.HelmChartProxy{}
+	key := types.NamespacedName{
+		Namespace: input.HelmChartProxy.Namespace,
+		Name:      input.HelmChartProxy.Name,
+	}
+	err = mgmtClient.Get(ctx, key, existing)
+	Expect(err).NotTo(HaveOccurred())
+
+	// Get the inital Helm release through the HelmChartProxy
```
Comment (mboersma) [inline test/e2e/helm_unchanged.go:72]: Same here--orphaned comment?

File: config/manager/manager.yaml [hunk 8]
```diff
@@ -33,6 +33,7 @@ spec:
           - --leader-elect
           - "--diagnostics-address=${CAAPH_DIAGNOSTICS_ADDRESS:=:8443}"
           - "--insecure-diagnostics=${CAAPH_INSECURE_DIAGNOSTICS:=false}"
+          - "--sync-period=${CAAPH_SYNC_PERIOD:=10m}"
```
Comment (mboersma) [inline config/manager/manager.yaml:36]: Is this change related to the rest of the PR?

**Review by mboersma — COMMENTED** [review]
This looks great, nice job on the tests in particular. I had a few small comments, but lgtm.

File: config/manager/manager.yaml (hunk 8 above)
Comment (Jont828) [reply config/manager/manager.yaml:36]: Yes, I added this so in the test, we could change the sync period to 1m, so that when we have the `time.Sleep()` after an out of band upgrade, we can be sure a reconciliation loop was triggered.

File: test/e2e/helm_test.go (hunk 6 above)
Comment (Jont828) [reply test/e2e/helm_test.go:339]: I just wanted to be sure that subsequent changes to the Helm chart would still be durable. I think it makes sense to put it in a loop, though.

File: test/e2e/helm_out_of_band.go (hunk 5 above)
Comment (Jont828) [reply test/e2e/helm_out_of_band.go]: I removed it as a TODO, but I think it's worth following up on if we later come up with a better idea on how to prove the Helm release won't change.

**Review by mboersma — COMMENTED** [review]
lgtm, needs squashing

**Comment by Jont828** [comment]
Done!

**Comment by Jont828** [comment]
@mboersma Can you LGTM?

**Review by mboersma — APPROVED** [review]
/lgtm
